linux/meshd-nl80211.c

```diff
@@ -1130,79 +1148,66 @@ void estab_peer_link(unsigned char *peer,
-{
-    /* Create a candidate */
-    struct nl_msg *msg;
-    uint8_t cmd = NL80211_CMD_NEW_STATION;
-    int ret;
-    char *pret;
-    uint8_t supported_rates[] = { 2, 4, 10, 22, 96, 108 }; /* XXX: Try to use IE rates, or otherwise basic rate? */
-    struct nl80211_sta_flag_update flags;
+        elems.sup_rates = cand->sup_rates;
+        elems.sup_rates_len = cand->sup_rates_len;
 
-    if (!peer)
-        return;
+        elems.ht_cap = (unsigned char *) &cand->ht_cap;
+        elems.ht_cap_len = sizeof(cand->ht_cap);
 
-    msg = nlmsg_alloc();
+        elems.ht_info = (unsigned char *) &cand->ht_info;
+        elems.ht_info_len = sizeof(cand->ht_info);
 
-    if (!msg)
-        return;
-
-    pret = genlmsg_put(msg, 0, 0,
-            genl_family_get_id(nlcfg.nl80211), 0, 0, cmd, 0);
+        ret = add_station(&nlcfg, peer, &elems);
+        if (ret)
+            return;
 
-    if (pret == NULL)
-        goto nla_put_failure;
+        cand->uploaded = true;
+    }
 
-    NLA_PUT_U32(msg, NL80211_ATTR_IFINDEX, nlcfg.ifindex);
-    NLA_PUT(msg, NL80211_ATTR_MAC, ETH_ALEN, peer);
-    NLA_PUT(msg, NL80211_ATTR_STA_SUPPORTED_RATES, sizeof(supported_rates),
-            supported_rates);
-    flags.mask = (1 << NL80211_STA_FLAG_AUTHENTICATED) |
-                 (1 << NL80211_STA_FLAG_WME);
-    flags.set = (1 << NL80211_STA_FLAG_WME);
-    NLA_PUT(msg, NL80211_ATTR_STA_FLAGS2, sizeof(flags), &flags);
+    set_authenticated_flag(&nlcfg, peer);
+    /* key to encrypt/decrypt unicast data AND mgmt traffic to/from this peer */
+	install_key(&nlcfg, peer, CIPHER_CCMP, NL80211_KEYTYPE_PAIRWISE, 0, mtk);
 
-    /* unused for mesh but mandatory for NL80211_CMD_NEW_STATION */
-    NLA_PUT_U16(msg, NL80211_ATTR_STA_AID, 1);
-    NLA_PUT_U16(msg, NL80211_ATTR_STA_LISTEN_INTERVAL, 100);
+    /* key to decrypt multicast data traffic from this peer */
+	install_key(&nlcfg, peer, CIPHER_CCMP, NL80211_KEYTYPE_GROUP, 1, peer_mgtk);
 
-    nlcfg.supress_error = -EEXIST;
-    ret = send_nlmsg(nlcfg.nl_sock, msg);
-    sae_debug(MESHD_DEBUG, "new peer candidate (seq num=%d)\n",
-            nlmsg_hdr(msg)->nlmsg_seq);
-    nlmsg_free(msg);
-    msg = NULL;
-    if (ret < 0)
-        fprintf(stderr,"New candidate failed: %d (%s)\n", ret, strerror(-ret));
-    else
-        ret = 0;
+    /* to check integrity of multicast mgmt frames from this peer */
+    if (peer_igtk) {
+	    install_key(&nlcfg, peer, CIPHER_AES_CMAC, NL80211_KEYTYPE_GROUP, peer_igtk_keyid, peer_igtk);
+    }
+}
 
-    return;
-nla_put_failure:
-    nlmsg_free(msg);
-    msg = NULL;
+void peer_created(unsigned char *peer)
+{
+    /*
+     * We don't have to do anything here:
+     *
+     *  - if peer was created from SAE state machine, then
+     *    it will be added to kernel when the IEs are known (e.g.
+     *    on estab plink)
+     *
+     *  - if peer is created from a beacon, we will know the IEs
+     *    and will go ahead and create the peer in new_candidate_handler.
+     */
```

Comment:
Is there anything we can check in the `peer` that would confirm that we are in one of the two scenarios described?  Maybe...
```
if (peer->uploaded) {
  sae_debug(MESHD_DEBUG_ERR, "SAE attempting to create a peer that is already in kernel\n");
}
```

Also, could you check if this is still accurate after your changes? https://github.com/cozybit/authsae/blob/6f5be621264357e30808ea354815efa3da38262c/linux/meshd-nl80211.c#L68